core/src/main/java/com/google/instrumentation/tags/TagSet.java

```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright 2017, Google Inc.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.instrumentation.tags;
+
+import com.google.instrumentation.internal.StringUtil;
+import javax.annotation.Nullable;
+
+/**
+ * A set of key-value pairs that can be used to label anything that is associated with a specific
+ * operation.
+ *
+ * <p>For example, {@code TagSet}s can be used to label stats, log messages, or debugging
+ * information.
+ */
+public abstract class TagSet {
+  /** The maximum length for a string tag value. */
+  public static final int MAX_STRING_LENGTH = StringUtil.MAX_LENGTH;
+
+  /**
+   * Determines whether a key is present.
+   *
+   * @param key the key to look up.
+   * @return {@code true} if the key is present.
+   */
+  public abstract boolean tagKeyExists(TagKey<?> key);
+
+  // TODO(sebright): Which is better, the generic "get" method or three specialized "get" methods?
+  // The specialized "get" methods can return primitives, which could be less expensive.
+
+  // TODO(sebright): Should these methods take defaults or throw exceptions to handle missing keys?
```

Comment:
The trouble with default values is that there may not be one - especially for bool. I think that leaves your only other option as being throwing; if users are want to avoid that, they can use tagKeyExists first.